Add cancellable return to http methods

danielCarlosCE opened the pull request:
Fixes #11

danielCarlosCE committed:
Add Cancellable return to Getable’s functions

  - Not sure if URLError.cancelled logic should go on ResultGenerator. It doesn’t seem right. Other options would be FrisbeeError and DataTaskRunner
  - NillCancellable is a Null Object pattern use
  - For some unknown reason, NetworkGet is only 95.12% coverage

danielCarlosCE commented:
I'm facing this exactly same issue with coverage
https://stackoverflow.com/questions/34622082/why-is-a-closing-brace-showing-no-code-coverage

ronanrodrigo commented on Sources/Frisbee/Interactors/ResultGenerator.swift:
```diff
@@ -10,8 +10,14 @@ final class ResultGenerator<T: Decodable> {
 
     func generate(data: Data?, error: Error?) -> Result<T> {
         guard let data = data else {
-            let frisbeeError = FrisbeeError(error ?? FrisbeeError.noData)
```
I was trying to abstract error construction inside the `FrisbeError` constructor.  What you think? If you agree, could be something like this:
```swift
public enum FrisbeeError: Error {

    // ...

    init(_ error: Error?) {
        guard let frisbeeError = error as? FrisbeeError else {
            switch error {
                 case .some(URLError.cancelled):
                     self = .requestCancelled
                 case .some(let error):
                     self = .other(localizedDescription: error.localizedDescription)
                 case .none:
                     return .noData
             }
        }
        self = frisbeeError
    }
}
```

With this, it isn't necessary to to change the `ResultGeneratorTests`. Just add an test for `FrisbeeError`

ronanrodrigo replied:
Maybe with my suggestion, the best approach is to replace `guard/let` by `if/let`

ronanrodrigo commented on Sources/Frisbee/Interactors/Cancellable.swift:
```diff
@@ -0,0 +1,7 @@
+    func cancel()
+}
+
+final class NilCancellable: Cancellable {
```
What you think about `Uncancellable` for this class name?

ronanrodrigo replied:
I was thinking and maybe this class is better to not exist. And instead `@discardableResults` in `get` functions it is better to use polimorphism overloading `get` function adding functions with `Cancelable` return.

danielCarlosCE replied:
@ronanrodrigo I am not sure we can use polymorphism in this case. I used the NilCancellable in cases of `onComplete(.fail`. The function returns a task when success, but in case of failure we just return a class that will do nothin on `cancel` call. 

The name I chose was to recall the pattern. It should be NullCancellable, actually, but since it's Swift. I think Uncancellable will bring ubiquity to its behavior.

ronanrodrigo replied:
Yes I understand your motivation and because of the `@discardableResult` who will use the Frisbee will not be affected by this decision. But an empty implementation of `cancel` method seems a code smell.

About the polymorphism, it is possible with we do that:
```swift
public final class NetworkGet: Getable {
    // ...
    public func get<T: Decodable, U: Encodable>(url: String, query: U, onComplete: @escaping OnComplete<T>) {
        guard let url = URL(string: url) else { return onComplete(.fail(.invalidUrl)) }
        return get(url: url, query: query, onComplete: onComplete) // 👈 it is necessary to use the "return" to avoid Ambiguous error
    }

    public func get<T: Decodable, U: Encodable>(url: URL, query: U, onComplete: @escaping OnComplete<T>) {
        // ...
        makeRequest(url: url, onComplete: onComplete)
    }

    public func get<T: Decodable, U: Encodable>(url: String, query: U, onComplete: @escaping OnComplete<T>) -> Cancelable {
        return makeRequest(url: url, onComplete: onComplete)
    }
}
```

ronanrodrigo commented on Sources/Frisbee/Interactors/Cancellable.swift:
```diff
@@ -0,0 +1,7 @@
+public protocol Cancellable {
```
To wrapping direct Foundation stuffs it is used Adapter pattern. It is exactly what you do, but in Frisbee we are using the Adapters folder and Adapter suffix. Could you update?

danielCarlosCE replied:
Would you call CancellableAdapter, URLSessionTaskAdapter, SessionTaskAdapter or something else?  I didn't see any adapter the Foundation class could just adopt so I wonder if I need to change something besides the name.

ronanrodrigo replied:
Just the names and folder location. The protocol will be `CancellableAdapter` and the implementation with Foudantation library will be `URLSessionTaskAdapter`.

ronanrodrigo reviewed (changes requested):
Very nice @danielCarlosCE! I just raised a few small points to update.

Thanks 👏😀

danielCarlosCE committed:
Turn Cancellable into CancellableAdapter

  - Move URLError.cancelled logic to FrisbeeError’s init
  - Create URLSessionTaskAdapter instead of using URLSessionTask extension

danielCarlosCE commented:
I added a new TestCase called URLSessionTaskAdapterTests. Should I add `static var allTests` to it? I'm not sure why you guys are using this, but there are some TestCases without it.

danielCarlosCE commented:
@ronanrodrigo I guess I followed the suggested names, but reflecting about it, I don't think Cancellable should be name Adapter. Only URLSessionTaskAdapter is actually an adapter. Would you agree?

ronanrodrigo commented:
> I added a new TestCase called URLSessionTaskAdapterTests. Should I add static var allTests to it? I'm not sure why you guys are using this, but there are some TestCases without it.

`allTests` is for running tests on linux environment and some tests without `allTests` aren't running in linux because of some particularities. You could leave without `allTests` and I will handle this.


> @ronanrodrigo I guess I followed the suggested names, but reflecting about it, I don't think Cancellable should be name Adapter. Only URLSessionTaskAdapter is actually an adapter. Would you agree?

You're right. Just `URLSessionTaskAdapter` is an adapter.

danielCarlosCE committed:
Turn CancellableAdapter back to Cancellable

  because only URSessionTask is a real “Adapter”
  - Make Postable return Cancellable
  - Make sure methods with url: String version just validates the string and call its sibling url: URL, to avoid repeating code

ronanrodrigo commented:
Thanks @danielCarlosCE 🎉😀
